### PR TITLE
CE-3440: Edit highlighted text

### PR DESCRIPTION
**Pull request by lukaszkonieczny**
## Links

* https://wikia-inc.atlassian.net/browse/CE-3429

## Description

When user highlights part of article text, the edit button appears. If user clicks on it, section editor is loaded with highlighted text.

## Reviewers

@Wikia/community-engineering @Wikia/x-wing

### CODE REVIEW

File: front/main/app/components/article-edit-highlighted-entrypoint.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import Ember from 'ember';
+
+export default Ember.Component.extend({
+	classNames: ['article-edit-highlighted-entrypoint'],
+	classNameBindings: ['displayEdit'],
+	displayEdit: Ember.computed('showEdit', function () {
```
Comment (hakubo) [inline front/main/app/components/article-edit-highlighted-entrypoint.js]: in templates you can use inline if
`class="{{if showEdit 'show' 'hide'}}-edit"`

File: front/main/app/components/article-wrapper.js [hunk 2]
```diff
@@ -73,6 +79,29 @@ export default Ember.Component.extend(
...
+				highlightedText = this.replaceTags(highlightedText);
+
+				this.setHighlightedTextVars(sectionIndex, highlightedText, true);
+			} else {
+				this.setHighlightedTextVars(0, '', false);
+			}
+		},
+
+		setHighlightedTextVars(sectionIndex, highlightedText, showEdit) {
+			this.set('highlightedSectionIndex', sectionIndex);
```
Comment (hakubo) [inline front/main/app/components/article-wrapper.js]: setProperties

File: front/main/app/templates/components/article-wrapper.hbs [hunk 3]
```diff
@@ -115,6 +115,13 @@
 		{{#if displayRecentEdit}}
 			{{recent-edit-banner}}
 		{{/if}}
+		{{article-edit-highlighted-entrypoint
+			section=highlightedSectionIndex
+			highlightedText=highlightedText
+			title=model.title
+			showEdit=showHighlightedEdit
```
Comment (hakubo) [inline front/main/app/templates/components/article-wrapper.hbs]: can this use closure action as well?

File: front/main/app/mixins/text-highlight.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	selection: null,
+
+	getSelection() {
+		const selection = this.get('selection');
+
+		return selection ? selection : window.getSelection();
+	},
+
+	setSelection(selection) {
+		this.set('selection', selection);
```
Comment (hakubo) [inline front/main/app/mixins/text-highlight.js]: why one should use that over just calling .set('selection')?

File: front/main/app/mixins/text-highlight.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+		const fragment = this.getSelection().getRangeAt(0).cloneContents(),
+			div = document.createElement('div');
+
+		div.appendChild(fragment.cloneNode(true));
+
+		return div.innerHTML;
+	},
+
+	removeTagsFromBegining(text) {
+		while (firstTagPattern.test(text)) {
```
Comment (hakubo) [inline front/main/app/mixins/text-highlight.js]: to not loop - can we use /g modifier for this regexp?

**Comment by hakubo** [comment]
my notes are optional; LGTM if you don't have time to work on that more

File: front/main/app/components/article-edit-highlighted-entrypoint.js (hunk 1 above)
Comment (lukaszkonieczny) [inline front/main/app/components/article-edit-highlighted-entrypoint.js]: I need third option here. Don't add any class until user highlight some text.

File: front/main/app/mixins/text-highlight.js (hunk 5 above)
Comment (lukaszkonieczny) [inline front/main/app/mixins/text-highlight.js]: while using "begining" and "end" string position I cannot use /g modifier. But I improved regexps do remove loops :) https://github.com/Wikia/mercury/commit/166b030e5a070a8be5cc9ec0f5105c5bc5a89d3e

File: front/main/app/mixins/text-highlight.js (hunk 4 above)
Comment (lukaszkonieczny) [inline front/main/app/mixins/text-highlight.js]: To make it more readable while it's public function and can be used in different places and it's not required to set this selection but using it give some advantages.

File: front/main/app/components/article-edit-highlighted-entrypoint.js (hunk 1 above)
Comment (Warkot) [inline front/main/app/components/article-edit-highlighted-entrypoint.js]: BTW, this inline method looks fancy, but grepping code like this is not easy.

File: front/main/app/components/article-edit-highlighted-entrypoint.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+		const showEdit = this.get('showEdit');
+
+		let className = '';
```
Comment (Warkot) [inline front/main/app/components/article-edit-highlighted-entrypoint.js]:
```
if (showEdit === true) {
	return 'show-edit';
} else if (showEdit === false) {
	return 'hide-edit';
}

return '';
```

File: front/main/app/symbols/main/pen.svg [hunk 7]
```diff
@@ -1,4 +1 @@
-<svg xmlns:svg="http://www.w3.org/2000/svg" xmlns="http://www.w3.org/2000/svg" version="1.1" x="0" y="0" viewBox="0 0 15.589231 15.78923" enable-background="new 0 0 40 40" xml:space="preserve" width="15.589231" height="15.78923">
-  <path d="m1.8392305 11.75c-0.1-0.1-0.3-0.1-0.4 0.1l-1.40000002 3.7c-0.1 0.2 0 0.3 0.2 0.2l3.70000002-1.4c0.2-0.1 0.2-0.2 0.1-0.4zM13.53923 0.15c-0.2-0.2-0.5-0.2-0.7 0l-0.3 0.4-1 1 0.3 0.3 2 2 0.3 0.3 1-1 0.3-0.3c0.2-0.2 0.2-0.5 0-0.7z" fill="#000"/>
-  <rect x="-12.948761" y="-5.8123183" transform="rotate(-134.9919)" width="3.7999635" height="11.799887" fill="#000"/>
-</svg>
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 15.589231 15.78923" width="15.589" height="15.789"><path d="M1.84 11.75c-.1-.1-.3-.1-.4.1l-1.4 3.7c-.1.2 0 .3.2.2l3.7-1.4c.2-.1.2-.2.1-.4zM13.54.15c-.2-.2-.5-.2-.7 0l-.3.4-1 1 .3.3 2 2 .3.3 1-1 .3-.3c.2-.2.2-.5 0-.7zM5.045 13.267L2.358 10.58l8.345-8.343 2.687 2.687z"/></svg>
```
Comment (Warkot) [inline front/main/app/symbols/main/pen.svg]: `viewBox="0 0 15.589231 15.78923"` do we need such precision? :wink:

File: front/main/app/components/article-edit-highlighted-entrypoint.js (hunk 1 above)
Comment (vforge) [inline front/main/app/components/article-edit-highlighted-entrypoint.js]: I support @warkot here - let's not concat class names / i18n keys when not necessary.

**Pull request merged**